quickwit-search/src/search_client_pool.rs

```diff
@@ -241,33 +247,44 @@ impl SearchClientPool {
             }
         }
 
-        // Sort job
-        jobs.sort_by(|left, right| {
+        // sort jobs by cost desc, split_id asc.
+        //
+        // By keeping split_id in comparision, we minimize cost for next rendez_vous sort when
+        // two consecutive split_id(s) are the same.
+        jobs.sort_unstable_by(|left, right| {
             // sort_by_key does not work here unfortunately
             job_order_key(left).cmp(&job_order_key(right))
         });
 
+        // Compute load (per node) threshold
+        let total_cost: Cost = jobs.iter().map(|v| v.cost()).sum();
+        let load_per_node_threshold =
+            ((total_cost as f64 / nodes.len() as f64) * (1.0 + LOAD_MARGIN)).ceil() as u64;
+        let load_per_node_threshold = load_per_node_threshold.max(MIN_LOAD_THRESHOLD);
+
         for job in jobs {
             sort_by_rendez_vous_hash(&mut nodes, job.split_id());
-            // choose one of the the first two nodes based on least loaded
-            let chosen_node_index: usize = if nodes.len() >= 2 {
-                if nodes[0].load > nodes[1].load {
-                    1
-                } else {
-                    0
-                }
-            } else {
-                0
-            };
 
             // update node load for next round
-            nodes[chosen_node_index].load += job.cost() as u64;
+            nodes[0].load += job.cost() as u64;
```

Comment:
The resulting load may be much higher than the load_per_node threshold.

Reply:
with the formula I suggest above, I think we need to add a check on the load not exceeding the threshold after the assignment here.